src/Core/Form/IdentifiableObject/Builder/FormBuilder.php

```diff
@@ -0,0 +1,129 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder;
+
+use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataProvider\FormDataProviderInterface;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
+
+/**
+ * Creates new forms for identifiable objects.
+ */
+final class FormBuilder implements FormBuilderInterface
+{
+    /**
+     * @var FormFactoryInterface
+     */
+    private $formFactory;
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @var FormDataProviderInterface
+     */
+    private $dataProvider;
+
+    /**
+     * @var string
+     */
+    private $formType;
+
+    /**
+     * @param FormFactoryInterface $formFactory
+     * @param HookDispatcherInterface $hookDispatcher
+     * @param FormDataProviderInterface $dataProvider
+     * @param $formType
+     */
+    public function __construct(
+        FormFactoryInterface $formFactory,
+        HookDispatcherInterface $hookDispatcher,
+        FormDataProviderInterface $dataProvider,
+        $formType
+    ) {
+        $this->formFactory = $formFactory;
+        $this->hookDispatcher = $hookDispatcher;
+        $this->dataProvider = $dataProvider;
+        $this->formType = $formType;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getForm(array $data = [], array $options = [])
+    {
+        if (is_array($defaultData = $this->dataProvider->getDefaultData())) {
+            $data = array_merge($defaultData, $data);
+        }
+
+        return $this->buildForm(
+            $this->formType,
+            $data,
+            null,
+            $options
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFormFor($id, array $data = [], array $options = [])
+    {
+        $data = array_merge($this->dataProvider->getData($id), $data);
+
+        return $this->buildForm(
+            $this->formType,
+            $data,
+            $id,
+            $options
+        );
+    }
+
+    /**
+     * @param string $formType
+     * @param array $data
+     * @param int|null $id
+     * @param array $options
+     *
+     * @return FormInterface
+     */
+    private function buildForm($formType, $data, $id = null, array $options = [])
+    {
+        $formBuilder = $this->formFactory->createBuilder($formType, $data, $options);
+
+        $this->hookDispatcher->dispatchWithParameters('action' . $formBuilder->getName() . 'FormBuilderModifier', [
```

Comment:
i agree with that, we could specify custom block prefix for some (or maybe all) types. Like `product_settings` or `customer_settings` so `SettingsType` would be unique as in your example.

Or we could even put fully qualified class name into hook name, like its done in object model.

Reply:
yep, the good thing is for now all our types seem to have overridden `getBlockPrefix` so they should be unique
Indeed using the fully qualified name would solve everything, but it's gonna create reaaaally long hook names, which is why I'm not fan of this solution

You said it's donne in ObjecModel, do you mean le legacy objects? do you have an example please?

Reply:
yup, in legacy object model, here's one of the example, where it uses FQN as hook name:

https://github.com/PrestaShop/PrestaShop/blob/develop/classes/ObjectModel.php#L530

but i agree with you, it will make hook names very long and thus its not really a good solution. :/